include/ginkgo/core/matrix/permutation.hpp

```diff
@@ -196,71 +269,34 @@ class Permutation : public EnableLinOp<Permutation<IndexType>>,
-          row_size_(size[0]),
-          col_size_(size[1]),
-          enabled_permute_(enabled_permute)
-    {
-        if (enabled_permute_ & row_permute) {
-            GKO_ASSERT_EQ(size[0], permutation_.get_num_elems());
-        }
-        if (enabled_permute_ & column_permute) {
-            GKO_ASSERT_EQ(size[1], permutation_.get_num_elems());
-        }
-    }
+    Permutation(std::shared_ptr<const Executor> exec,
+                array<index_type> permutation_indices);
 
-    void apply_impl(const LinOp* in, LinOp* out) const
-    {
-        auto perm = as<Permutable<index_type>>(in);
-        std::unique_ptr<gko::LinOp> tmp{};
-        if (enabled_permute_ & inverse_permute) {
-            if (enabled_permute_ & row_permute) {
-                tmp = perm->inverse_row_permute(&permutation_);
-            }
-            if (enabled_permute_ & column_permute) {
-                if (enabled_permute_ & row_permute) {
-                    tmp = as<Permutable<index_type>>(tmp.get())
-                              ->inverse_column_permute(&permutation_);
-                } else {
-                    tmp = perm->inverse_column_permute(&permutation_);
-                }
-            }
-        } else {
-            if (enabled_permute_ & row_permute) {
-                tmp = perm->row_permute(&permutation_);
-            }
-            if (enabled_permute_ & column_permute) {
-                if (enabled_permute_ & row_permute) {
-                    tmp = as<Permutable<index_type>>(tmp.get())->column_permute(
-                        &permutation_);
-                } else {
-                    tmp = perm->column_permute(&permutation_);
-                }
-            }
-        }
-        out->move_from(tmp);
-    }
+    [[deprecated(
+        "dim<2> is no longer supported as a dimension parameter, use size_type "
+        "instead")]] Permutation(std::shared_ptr<const Executor> exec,
+                                 const dim<2>& size);
 
+    [[deprecated("permute mask is no longer supported")]] Permutation(
+        std::shared_ptr<const Executor> exec, const dim<2>& size,
+        const mask_type& enabled_permute);
 
-    void apply_impl(const LinOp*, const LinOp* in, const LinOp*,
-                    LinOp* out) const
-    {
-        // Ignores alpha and beta and just performs a normal permutation as an
-        // advanced apply does not really make sense here.
-        this->apply_impl(in, out);
-    }
+    [[deprecated("use the overload without dimensions")]] Permutation(
+        std::shared_ptr<const Executor> exec, const dim<2>& size,
+        array<index_type> permutation_indices);
 
+    [[deprecated("permute mask is no longer supported")]] Permutation(
+        std::shared_ptr<const Executor> exec, const dim<2>& size,
+        array<index_type> permutation_indices,
+        const mask_type& enabled_permute);
+
+    void apply_impl(const LinOp* in, LinOp* out) const override;
```

Comment:
Yes. The only operations that make sense as an `apply` are row operations (permutation, inverse permutation), otherwise the dimensions no longer conform for rectangular inputs. For `A->permute(P)`, I wanted an object that imbues an `array<int>` with meaning.